pkg/nsx/services/vpc/vpc.go

```diff
@@ -471,3 +514,230 @@ func (s *VPCService) Cleanup() error {
+	key := "Tag"
+	expression := service.buildExpression(resource_type, member_type, value, key, operator, scope_op)
+	group.Expression = []*data.StructValue{expression}
+	group.DisplayName = pointy.String(groupId)
+
+	err := service.NSXClient.VpcGroupClient.Patch(orgId, projectId, vpcId, groupId, group)
+	if err != nil {
+		return &group, err
+	}
+	nsxgroup, err := service.NSXClient.VpcGroupClient.Get(orgId, projectId, vpcId, groupId)
+	return &nsxgroup, err
+}
+
+func (service *VPCService) buildExpression(resource_type, member_type, value, key, operator, scope_op string) *data.StructValue {
+	return data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type":  data.NewStringValue(resource_type),
+			"member_type":    data.NewStringValue(member_type),
+			"value":          data.NewStringValue(value),
+			"key":            data.NewStringValue(key),
+			"operator":       data.NewStringValue(operator),
+			"scope_operator": data.NewStringValue(scope_op),
+		},
+	)
+}
+
+func (service *VPCService) buildAVIAllowRule(obj *model.Vpc, externalCIDR []string, groupId, ruleId, projectId string) (*model.Rule, error) {
+	rule := &model.Rule{}
+	rule.Action = pointy.String(model.Rule_ACTION_ALLOW)
+	rule.Direction = pointy.String(model.Rule_DIRECTION_IN_OUT)
+	rule.Scope = append(rule.Scope, groupId)
+	rule.SequenceNumber = pointy.Int64(math.MaxInt32 - 1)
+	rule.DestinationGroups = externalCIDR
+	rule.SourceGroups = append(rule.SourceGroups, "Any")
+	name := fmt.Sprintf("PROJECT-%s-VPC-%s-%s", projectId, *obj.Id, ruleId)
+	rule.DisplayName = pointy.String(name)
+	rule.Id = pointy.String(ruleId)
+	rule.Services = []string{"ANY"}
+	rule.IsDefault = pointy.Bool(true)
+	tags := []model.Tag{
+		{
+			Scope: pointy.String(common.TagScopeCluster),
+			Tag:   pointy.String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVersion),
+			Tag:   pointy.String(strings.Join(common.TagValueVersion, ".")),
+		},
+	}
+	rule.Tags = tags
+	return rule, nil
+}
+
+func (service *VPCService) getAVIAllowRule(orgId string, projectId string, vpcId string, spId string, ruleId string) (*model.Rule, error) {
+	key := fmt.Sprintf("/orgs/%s/projects/%s/vpcs/%s/security-policies/%s/rules/%s", orgId, projectId, vpcId, spId, ruleId)
+	rule := service.RuleStore.GetByKey(key)
+	if rule == nil {
+		log.Info("avi rule not found", "key", key)
+		return nil, errors.New("avi rule not found")
+	}
+	return rule, nil
+}
+
+func (service *VPCService) getAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	key := fmt.Sprintf("/orgs/%s/projects/%s/vpcs/%s/groups/%s", orgId, projectId, vpcId, groupId)
+	group := service.GroupStore.GetByKey(key)
+	var err error
+	if group == nil {
+		log.Info("avi group not found", "key", key)
+		err = errors.New("avi group not found")
+	}
+	return group, err
+}
+
+func (service *VPCService) existAVISecurityPolicy(orgId string, projectId string, vpcId string, spId string) bool {
```

Comment:
could we change func name to checkAVISecurityPolicyExist or something similar?
Also add comments above the func.
// checkAVISecurityPolicyExist returns true if namespace already exists

Reply:
done